pkg/segment/aggregations/generateevents.go

```diff
@@ -86,3 +99,112 @@ func PerformGenTimes(aggs *structs.QueryAggregators) error {
 
 	return nil
 }
+
+func checkCSVFormat(filename string) bool {
+	return strings.HasSuffix(filename, ".csv") || strings.HasSuffix(filename, ".csv.gz")
+}
+
+// Ensure record and column lengths are valid before calling this function.
+func createRecord(columnNames []string, record []string) map[string]interface{} {
+	recordMap := make(map[string]interface{})
+	for i, col := range columnNames {
+		recordMap[col] = record[i]
+	}
+	return recordMap
+}
+
+func PerformInputLookup(aggs *structs.QueryAggregators) error {
+	var reader *csv.Reader
+	if aggs.GenerateEvent.InputLookup == nil {
+		return fmt.Errorf("PerformInputLookup: InputLookup is nil")
+	}
+	filename := aggs.GenerateEvent.InputLookup.Filename
+
+	if !checkCSVFormat(filename) {
+		return fmt.Errorf("PerformInputLookup: Only CSV format is currently supported")
+	}
+
+	workingDir, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("PerformInputLookup: Error while getting current working directory, err: %v", err)
+	}
+	filepath := workingDir + "/" + config.GetLookupPath() + filename
+
+	file, err := os.Open(filepath)
+	if err != nil {
+		return fmt.Errorf("PerformInputLookup: Error while opening file %v, err: %v", filepath, err)
+	}
+	defer file.Close()
+
+	if strings.HasSuffix(filename, ".csv.gz") {
+		gzipReader, err := gzip.NewReader(file)
+		if err != nil {
+			return fmt.Errorf("PerformInputLookup: Error while creating gzip reader, err: %v", err)
+		}
+		defer gzipReader.Close()
+		reader = csv.NewReader(gzipReader)
+	} else {
+		reader = csv.NewReader(file)
+	}
+
+	// read columns from first row of csv file
+	columnNames, err := reader.Read()
+	if err != nil {
+		return fmt.Errorf("PerformInputLookup: Error reading column names, err: %v", err)
+	}
+
+	curr := 0
+	for curr < int(aggs.GenerateEvent.InputLookup.Start) {
+		_, err := reader.Read()
+		if err != nil {
+			return fmt.Errorf("PerformInputLookup: Error skipping rows, err: %v", err)
+		}
+		curr++
+	}
+
+	InitGenEvent(aggs)
+	records := aggs.GenerateEvent.GeneratedRecords
+	recordsIndex := aggs.GenerateEvent.GeneratedRecordsIndex
+
+	SetGeneratedCols(aggs.GenerateEvent, columnNames)
+
+	key := 0
+	count := 0
+	fieldToValue := make(map[string]utils.CValueEnclosure)
+	for count < int(aggs.GenerateEvent.InputLookup.Max) {
+		recordKey := fmt.Sprintf("%v_%v", aggs.GenerateEvent.EventPosition, key)
+		csvRecord, err := reader.Read()
+		if err != nil {
+			// Check if we've reached the end of the file
+			if err.Error() == "EOF" {
+				break
+			}
+			return fmt.Errorf("PerformInputLookup: Error reading record, err: %v", err)
+		}
+
+		if len(csvRecord) != len(columnNames) {
+			return fmt.Errorf("PerformInputLookup: Error reading record, column count mismatch")
+		}
+
+		count++
+		record := createRecord(columnNames, csvRecord)
```

Comment:
nit: I personally find it really odd to increment `count` (the loop counter) in the middle of the loop instead of at the beginning or the end of the loop. I'd suggest moving it to the beginning since you have a `continue` lower

Reply:
addressed in the latest commit.